src/wallet/dump.cpp

```diff
@@ -78,3 +78,167 @@ bool DumpWallet(std::shared_ptr<CWallet> wallet, bilingual_str& error)
+    std::unique_ptr<DatabaseBatch> batch = db.MakeBatch();
+    batch->TxnBegin();
+
+    // Compute the checksum
+    CHashWriter hasher(0, 0);
+    uint256 checksum;
+
+    // Read the records from the dump file and write them to the database
+    bool ret = true;
+    bool first = true;
+    while (dump_file.good()) {
+        std::string key;
+        std::getline(dump_file, key, ',');
+        std::string value;
+        std::getline(dump_file, value, '\n');
+
+        if (key == "checksum") {
+            std::vector<unsigned char> parsed_checksum = ParseHex(value);
+            std::copy(parsed_checksum.begin(), parsed_checksum.end(), checksum.begin());
+            break;
+        }
+
+        std::string line = strprintf("%s,%s\n", key, value);
+        hasher.write(line.data(), line.size());
+
+        if (key.empty() || value.empty()) {
+            continue;
+        }
+
+        // Check magic for first record
+        if (first) {
+            if (key != DUMP_MAGIC) {
+                error = strprintf(_("Error: Dumpfile identifier record is incorrect. Got \"%s\", expected \"%s\"."), key, DUMP_MAGIC);
+                ret = false;
+                break;
+            }
+            // Check the version number (value of first record)
+            uint32_t ver;
+            if (!ParseUInt32(value, &ver)) {
+                error =strprintf(_("Error: Unable to parse version %u as a uint32_t"), value);
+                ret = false;
+                break;
+            }
+            if (ver != DUMP_VERSION) {
+                error = strprintf(_("Error: Dumpfile version is not supported. This version of bitcoin-wallet only supports version 1 dumpfiles. Got dumpfile with version %s"), value);
+                ret = false;
+                break;
+            }
+            first = false;
+            continue;
+        }
+
+        if (!IsHex(key)) {
+            error = strprintf(_("Error: Got key that was not hex: %s"), key);
+            ret = false;
+            break;
+        }
+        if (!IsHex(value)) {
+            error = strprintf(_("Error: Got value that was not hex: %s"), value);
+            ret = false;
+            break;
+        }
+
+        std::vector<unsigned char> k = ParseHex(key);
+        std::vector<unsigned char> v = ParseHex(value);
+
+        CDataStream ss_key(k, SER_DISK, CLIENT_VERSION);
+        CDataStream ss_value(v, SER_DISK, CLIENT_VERSION);
+
+        batch->Write(ss_key, ss_value);
+    }
+
+    if (ret) {
+        uint256 comp_checksum = hasher.GetHash();
+        if (checksum.IsNull()) {
+            error = _("Error: Missing checksum");
+            ret = false;
+        } else if (checksum != comp_checksum) {
+            error = strprintf(_("Error: Dumpfile checksum does not match. Computed %s, expected %s"), HexStr(comp_checksum), HexStr(checksum));
+            ret = false;
+        }
+    }
+
+    if (ret) {
+        batch->TxnCommit();
+    } else {
+        batch->TxnAbort();
```

Comment:
In commit "wallettool: Add createfromdump command" (6137750082dd28bf38bbb2922d6d040d4e854643)

Ideally, it would probably be preferable to delete the half created wallet than to just leave it behind. 

I think you could add `if (!ret) wallet->Delete()` after `Wallet->Close()` below with Delete method implementations that just delete the main datafile after the database is closed.

Reply:
It would be preferable to do that but deleting things was non-trivial enough that I didn't want to spend the effort to figure it out.

Reply:
> It would be preferable to do that but deleting things was non-trivial enough that I didn't want to spend the effort to figure it out.

Sure. I don't know, but I'd expect a simple implementation like the following to be adequate and safe:

```c++
assert(!m_db);
fs::remove(m_file_path);
fs::remove(m_dir_path);
```

Reply:
Added wallet removal.